lldb/source/Plugins/Process/elf-core/ProcessElfGpuCore.cpp

```diff
@@ -0,0 +1,165 @@
+
+lldb::ProcessSP ProcessElfGpuCore::CreateInstance(lldb::TargetSP target_sp,
+                                                  lldb::ListenerSP listener_sp,
+                                                  const FileSpec *crash_file,
+                                                  bool can_connect) {
+  return ProcessElfCore::CreateInstance(target_sp, listener_sp, crash_file,
+                                        can_connect);
+}
+
+lldb::ProcessSP ProcessElfGpuCore::FindGpuPlugin(
+    lldb::TargetSP target_sp, lldb::ListenerSP listener_sp,
+    const FileSpec *crash_file_path,
+    std::shared_ptr<ProcessElfCore> cpu_core_process) {
+  Log *log = GetLog(LLDBLog::Process);
+
+  ProcessSP process_sp;
+  ProcessCreateInstance create_callback = nullptr;
+
+  // Iterate through all registered GPU Process plugins
+  for (uint32_t idx = 0;
+       (create_callback =
+            PluginManager::GetGpuProcessCreateCallbackAtIndex(idx)) != nullptr;
+       ++idx) {
+
+    llvm::StringRef plugin_name =
+        PluginManager::GetGpuProcessPluginNameAtIndex(idx);
+    LLDB_LOGF(log, "FindGpuPlugin() - Trying GPU plugin: %s",
+              plugin_name.str().c_str());
+
+    // Try to create GPU process with this plugin
+    process_sp =
+        create_callback(target_sp, listener_sp, crash_file_path, false);
+
+    if (process_sp) {
+      // Set the CPU process BEFORE calling CanDebug so it has access to it.
+      // TODO: if we can extend ProcessCreateInstance to accept cpu process
+      // parameter this can be removed.
+      std::shared_ptr<ProcessElfGpuCore> gpu_process_sp =
+          std::static_pointer_cast<ProcessElfGpuCore>(process_sp);
+      if (gpu_process_sp) {
+        gpu_process_sp->SetCpuProcess(cpu_core_process);
+      }
+
+      // Now CanDebug can access the CPU process if needed
+      if (process_sp->CanDebug(target_sp, false)) {
+        LLDB_LOGF(log, "FindGpuPlugin() - Plugin %s accepted core",
+                  plugin_name.str().c_str());
+        break;
+      } else {
+        process_sp.reset();
+      }
+    }
+
+    LLDB_LOGF(log, "FindGpuPlugin() - Plugin %s declined core",
+              plugin_name.str().c_str());
+  }
+
+  return process_sp;
+}
+
+llvm::Expected<std::shared_ptr<ProcessElfGpuCore>>
+ProcessElfGpuCore::LoadGpuCore(std::shared_ptr<ProcessElfCore> cpu_core_process,
+                               const FileSpec &core_file) {
+  Log *log = GetLog(LLDBLog::Process);
+  LLDB_LOGF(
+      log,
+      "ProcessElfGpuCore::LoadGpuCore() - Looking for GPU data in core file");
+
+  lldb_private::Debugger &debugger =
+      cpu_core_process->GetTarget().GetDebugger();
+
+  // Create target for GPU
+  TargetSP gpu_target_sp;
+  llvm::StringRef exe_path;
+  llvm::StringRef triple;
+  OptionGroupPlatform *platform_options = nullptr;
+
+  Status error(debugger.GetTargetList().CreateTarget(
+      debugger, exe_path, triple, eLoadDependentsNo, platform_options,
+      gpu_target_sp));
```

Comment:
Each embedded core file plugin's `CreateInstance` function should be doing this work and each one should either return an empty ProcessSP if there there is no data in the `ProcessELFCore *native_core` argument, or creating its own target and process in that plug-in specific function.

Reply:
My intention is sharing GPU target creation code across all GPU vendors, but I can inline this into the process creation function if you want.